plugins/aggregators/derivative/README.md

```diff
@@ -0,0 +1,72 @@
+# Derivative Aggregator Plugin
+The Derivative Aggregator Plugin estimates the derivative for all fields of the
+aggregated metrics.
+
+### Time Derivatives
+
+In its default configuration it determines the first and last measurement of
+the period. From these measurements the time difference in seconds is
+calculated. This time difference is than used to divide the difference of each
+field using the following formula:
+```
+              field_last - field_first
+derivative = --------------------------
+                  time_difference
+```
+For each field the derivative is emitted with a naming pattern
+`fieldname_by_seconds`.
+
+### Custom Derivation Variable
+
+The plugin supports to use a field of the aggregated measurements as derivation
+variable in the denominator. This variable is assumed to be a monotonously
+increasing value. In this feature the following formula is used:
+```
+                 field_last - field_first
+derivative = --------------------------------
+              variable_last - variable_first
+```
+For each field the derivative then is emitted with a naming pattern
+`fieldname_by_variablename`.
+
+### Roll-Over to next Period
+
+By default the last measurement is used as first measurement in the next
+aggregation period. This enables a continuous calculation of the derivative. If
+within the next period an earlier timestamp is encountered this measurement will
+replace the roll-over metric. A main benefit of this roll-over is the ability to
+cope with multiple "quiet" periods, where no new measurement is pushed to the
+aggregator. The roll-over will take place at most `max_roll_over`times.
+
+### Configuration
+
+```toml
+[[aggregators.derivative]]
+  ## Specific Derivative Aggregator Arguments:
+  ## Configure a custom derivation variable. Timestamp is used if none is given.
+  # variable = "fieldname_of_derivation_vaiable"
+  ## Infix to separate field name from derivation variable.
+  infix = "_by_"  # default: _by_
+  ## Roll-Over last measurement to first measuremen of next period
+  max_roll_over = 10 # default: 10
+
+  ## General Aggregator Arguments:
+  ## calculate derivative every 30 seconds
+  period = "30s"
+  ## fields for which the derivative should be calculated
+  ## Important: The derivation variable must be contained in that listed
+  fieldpass = ["field1", "field2", "variable"]
```

Comment:
This collides with the generic telegraf `fieldpass` option. Please drop this option here.

Reply:
These comments were included to point out the necessity to include the derivation variable in the `fieldpass` argument. If it was not contained, it would not be accessible to the aggregator. The calculation could not be done in that case. How should we go about that?

Reply:
Well then please add a big fat note somewhere, otherwise it is easy for the reader to confuse `fieldspass` with a plugin-option. IIRC, you also issue an error in the plugin if this is not the case.

So my solution would be to add a **bold** note to the readme stating that if you use custom divisor variables those **have to be present** in case of field filtering. If you also want to put it to the config add `The specified variable field has to be present even is case of field filtering (fieldpass/fielddrop) or the plugin will fail!` to the description of the `variable` option.

Reply:
How about introducing a new config parameter `fields` just like in histogram or value_counter? That would remove the side-effect on `fieldpass` and ease the documentation.

Reply:
I don't like that approach of duplicating the `fieldpass` functionality. Please just add the documentation (which should be obvious for anybody anyway). The case of someone using a custom field (should be rare anyway) and then filtering that field out is clearly not a very common use-case either. Just make sure you catch that as an error and write a unit test for it.

Reply:
Can I get the `fieldpass` configuration injected into the derivative aggregator with something like this?
```go
type Derivative struct {
	FieldPass    string[]          `toml:"fieldpass"`
	Variable    string          `toml:"variable"`
        // ...
}
```

I would need this to check in `Init` whether `Variable` is contained as well.

Reply:
I think you shouldn't. Please simply check whether `Variable` is available in the metrics during `calculateDenominator()` or not and issue a log warn/error? if not.

Reply:
Could you please either remove the `fieldpass` option in the example or add a comment and link to the global telegraf option to clearly state that it is not specific to your aggregator!

Reply:
I removed `fieldpass` from the example.